pricegraph/src/orderbook.rs

```diff
@@ -140,6 +141,69 @@ impl Orderbook {
         debug_assert!(!self.is_overlapping());
     }
 
+    /// Reduces all overlapping orders for a given market and returns the
+    /// computed overlapping transitive orderbook for that market.
+    pub fn reduce_overlapping_transitive_orderbook(
+        &mut self,
+        base: TokenId,
+        quote: TokenId,
+    ) -> TransitiveOrderbook {
+        let (base, quote) = (node_index(base), node_index(quote));
+
+        let mut overlap = TransitiveOrderbook::default();
+        while let Err(NegativeCycle(predecessors, node)) =
+            bellman_ford::search(&self.projection, quote)
+        {
+            let path = {
+                let mut cycle = path::find_cycle(&predecessors, node, Some(quote))
+                    .expect("negative cycle not found after being detected");
+                cycle.push(cycle[0]);
+                cycle
+            };
+
+            if path[0] == quote {
```

Comment:
ubernit: We now paths is not empty but it's technically not checked anywhere. Maybe using first here is "safer"

Reply:
I like using `first` better here as well. It relies on less implicit assumptions about the path, and the compiler should be smart enough to optimize this away anyways.